Add new `contemporary` style

jalopezg-git opened the pull request:
This pull request introduces the new `contemporary` theme, largely based on `classic`, yet looking more up-to-date (see below for a screenshot).  `contemporary` features a nice-looking header, incl. a QR-code pointing to `\@homepage`, vertical timelines that span over a section, and the possibility to add symbols (e.g. using `fontawesome5`) to the section names.

@xdanaux, @moderncv, Please review, and let me know whether changes / fixes are needed before merging.

### Changes
- Add `moderncvverticaltimeline.sty`, implementing typesetting of a vertical timeline using PGF/TikZ.
- Add a new color theme, cerulean, made specifically for the contemporary style.
- Add style `moderncvstylecontemporary.sty`, together with new head and body styles.

![contemporary-10pt](https://github.com/xdanaux/moderncv/assets/36541918/81f34c88-59db-4fca-b4ac-9939ddb31eb5)

mefuller commented:
Visually, this looks very nice. I will try to get to a proper review within the next few days

jalopezg-git commented:
FWIW, an example of a final end-result of my own résumé is available at https://jalopezg.dev/public/resume_en-US.pdf.
IMO, the aesthetics are more aligned with 2023 :slightly_smiling_face:!

gerardbosch commented:
Hi @jalopezg-git, just to inform you that in your real resume here https://jalopezg.dev/public/resume_en-US.pdf, the href in your website is missing the `:`, so the link is broken. There's `https//jalopezg.dev/` instead of `https//jalopezg.dev/`. Cheers!

jalopezg-git commented:
> Hi @jalopezg-git, just to inform you that in your real resume here https://jalopezg.dev/public/resume_en-US.pdf, the href in your website is missing the `:`, so the link is broken. There's `https//jalopezg.dev/` instead of `https//jalopezg.dev/`. Cheers!

Ah, many thanks for noticing, @gerardbosch!  I will upload a new version soon with a couple of more fixes :slightly_smiling_face:.

gerardbosch commented:
> > Hi @jalopezg-git, just to inform you that in your real resume here https://jalopezg.dev/public/resume_en-US.pdf, the href in your website is missing the `:`, so the link is broken. There's `https//jalopezg.dev/` instead of `https//jalopezg.dev/`. Cheers!
> 
> Ah, many thanks for noticing, @gerardbosch! I will upload a new version soon with a couple of more fixes 🙂.

You're welcome! Cool theme BTW 👏👏

jalopezg-git commented:
Happy New 2024, everybody :tada:!  I take the chance to remind that this PR is still missing a first review :slightly_smiling_face:!

LLdaniel commented on moderncvbodyvi.sty:
```diff
@@ -0,0 +1,180 @@
+    \ifthenelse{\equal{#4}{}}{}{, {\slshape#4}}%
+    \ifthenelse{\equal{#5}{}}{}{, #5}%
+    \ifthenelse{\equal{#6}{}}{}{, #6}%
+    .\strut%
```
Is the `.` needed here? Maybe I am looking over something.

jalopezg-git replied:
This has been inherited from the parent style (classic) - see `moderncvbodyi.sty` - ofc, we could also remove it :slightly_smiling_face:!

jalopezg-git replied:
Removed (it seems I based the style on an older version of `moderncvbodyi.sty`).

LLdaniel commented on moderncvbodyvi.sty:
```diff
@@ -0,0 +1,180 @@
+%                required packages
+%-------------------------------------------------------------------------------
+\RequirePackage{moderncvverticaltimeline}
+
```
Maybe we can also support the `moderncvskillmatrix`
```latex
\RequirePackage[moderncvbodyvi]{moderncvskillmatrix}
```

LLdaniel commented on moderncvbodyvi.sty:
```diff
@@ -0,0 +1,180 @@
+  \setlength{\doubleitemcolumnwidth}{\maincolumnwidth-\hintscolumnwidth-\separatorcolumnwidth-\separatorcolumnwidth}%
+  \setlength{\doubleitemcolumnwidth}{0.5\doubleitemcolumnwidth}%
+  \setlength{\listdoubleitemcolumnwidth}{\maincolumnwidth-\listitemsymbolwidth-\separatorcolumnwidth-\listitemsymbolwidth}%
+  \setlength{\listdoubleitemcolumnwidth}{0.5\listdoubleitemcolumnwidth}%
+  % regular lengths
+  \setlength{\parskip}{0\p@}}
```
In case we add the skill matrix to your new body here it would be necessary to recompute the lengths accordingly like this (between l. 63 and 64):
```latex
%\recompute@cvskillmatrix@lengths
```

The problem with the skill matrix here is (s. also `\ReguirePackage` command in  my above comment):
* including the skill matrix **without** recomputation of the lengths results in a broken skill matrix
* including the skill matrix **with** recomputation of the lengths destroys your pretty "contemporary" style

How did you manage to get such a nice skill matrix in your "real world" example you shared? Maybe we have to change `moderncvskillmatrix.sty` for the "contemporary" style. 

Note: My experimentation is based on the current `template.tex` in branch `master`

jalopezg-git replied:
Thanks for pointing at this!  I have clarified this aspect [here](https://github.com/moderncv/moderncv/pull/173#issuecomment-2079810006).

jalopezg-git replied:
Thank you for the review!  That is fixed now - in a follow-up PR we might take care of the "short" format for the skill matrix, i.e. only skill name and rating as in my real-world example.

LLdaniel commented on moderncvbodyvi.sty:
```diff
@@ -0,0 +1,180 @@
+%% start of file `moderncvbodyvi.sty'.
```
This is a comment in general and does not belong directly to `moderncvbodyvi`. I just did not know how to comment on a file that is not involved in the "files changed" tab ^^. Sry.

It would be nice if we can include your style into the `template.tex` file in order to get a first impression how to use new style. 
In principle I could also add this because it appears to me that there is not really much to do. My suggestions regarding changes for the template are like this:
```diff
diff --git a/template.tex b/template.tex
index 750b4ce..2d9b2a2 100644
--- a/template.tex
+++ b/template.tex
@@ -9,8 +9,8 @@
 \documentclass[11pt,a4paper,sans]{moderncv}        % possible options include font size ('10pt', '11pt' and '12pt'), paper size ('a4paper', 'letterpaper', 'a5paper', 'legalpaper', 'executivepaper' and 'landscape') and font family ('sans' and 'roman')
 
 % moderncv themes
-\moderncvstyle{classic}                            % style options are 'casual' (default), 'classic', 'banking', 'oldstyle' and 'fancy'
-\moderncvcolor{blue}                               % color options 'black', 'blue' (default), 'burgundy', 'green', 'grey', 'orange', 'purple' and 'red'
+\moderncvstyle{contemporary}                       % style options are 'casual' (default), 'classic', 'banking', 'oldstyle', 'fancy' and 'contemporary'
+\moderncvcolor{cerulean}                           % color options 'black', 'blue' (default), 'burgundy', 'green', 'grey', 'orange', 'purple' and 'red'; for 'contemporary' style use 'cerulean'
 %\renewcommand{\familydefault}{\sfdefault}         % to set the default font; use '\sfdefault' for the default sans serif font, '\rmdefault' for the default roman one, or any tex font name
 %\nopagenumbers{}                                  % uncomment to suppress automatic page numbering for CVs longer than one page
 
@@ -114,7 +114,7 @@
 %-----       resume       ---------------------------------------------------------
 \makecvtitle
 
-\section{Education}
+\section{Education}  % for 'contemporary' style use optional argument for displaying an icon: \section[<icon>]{Education}
 \cventry{year--year}{Degree}{Institution}{City}{\textit{Grade}}{Description}  % arguments 3 to 6 can be left empty
 \cventry{year--year}{Degree}{Institution}{City}{\textit{Grade}}{Description}
 
@@ -147,7 +147,7 @@ Detailed achievements:
 \subsection{Miscellaneous}
 \cventry{year--year}{Job title}{Employer}{City}{}{Description}
 
-\section{Languages}
+\section[\emailsymbol]{Languages}
 \cvitemwithcomment{Language 1}{Skill level}{Comment}
 \cvitemwithcomment{Language 2}{Skill level}{Comment}
 \cvitemwithcomment{Language 3}{Skill level}{Comment}
```
Once this PR is merged I also can try to add a more detailed explanation in the new user guide.

jalopezg-git replied:
Thanks for the suggestion.  Typically, the theme also looks better when the margins (esp. the top one) are reduced.  I can mention that also in `template.tex`.

jalopezg-git replied:
I have carried out these changes.  However, the theme looks kind-of broken due to the big heading; perhaps we could comment out by default some of the `\social` commands in the template (?).

Also, per the suggested changes above, `contemporary` would become the new default - I'm quite proud of that, haha!

LLdaniel reviewed (changes requested):
Lovely style, thanks for that contribution! 
All in all it looks very good to me. The main comments are just minor changes.

Although I had one issues when testing with the current `template.tex` regarding the skill matrix. Therefore I just wanted to ask you how you did manage to create your nice skill matrix in your real example?
 
Maybe we can try to find a solution to include the current skill matrix in your style as well. (see comments).
We could also try to address this issue in a separate issue later and remove my code line suggestions regarding the skill matrix. Then we should add a short note in the `template.tex` that one has to comment out all skill matrix lines.

jalopezg-git commented:
> Lovely style, thanks for that contribution! All in all it looks very good to me. The main comments are just minor changes.
> 

Many thanks for the review!  I'll try to address your comments in the coming days.

> Although I had one issues when testing with the current `template.tex` regarding the skill matrix. Therefore I just wanted to ask you how you did manage to create your nice skill matrix in your real example?
> 
> Maybe we can try to find a solution to include the current skill matrix in your style as well. (see comments). We could also try to address this issue in a separate issue later and remove my code line suggestions regarding the skill matrix. Then we should add a short note in the `template.tex` that one has to comment out all skill matrix lines.

For this to work, I basically hacked an ad-hoc replacement for some of the skill matrix macros (see attached file).  I can try to make moderncvskillmatrix officially work in this PR :slightly_frowning_face: - if I don't find the time, are you okay if we leave it as a follow-up?

[moderncvskillmatrixtweaks.tex.gz](https://github.com/moderncv/moderncv/files/15133715/moderncvskillmatrixtweaks.tex.gz)

LLdaniel commented:
> For this to work, I basically hacked an ad-hoc replacement for some of the skill matrix macros (see attached file). I can try to make moderncvskillmatrix officially work in this PR 🙂 - if I don't find the time, are you okay if we leave it as a follow-up?
>
> Attachment: [moderncvskillmatrixtweaks.tex.gz](https://github.com/moderncv/moderncv/files/15133715/moderncvskillmatrixtweaks.tex.gz)

Thanks for your attachment. First I just thought we could try to support the examples in the `templates.tex`. But if there is more work to be done to integrate the skill matrix maybe  we can do that in a separate task, yes.

jalopezg-git commented on moderncvbodyvi.sty:
```diff
@@ -0,0 +1,180 @@
+  \setlength{\parskip}{6\p@}}
+
+\renewcommand*{\makeletterclosing}{
+  \@closing\\[3em]%
```
Done!

jalopezg-git commented on moderncvheadvii.sty:
```diff
@@ -0,0 +1,223 @@
+  \hfill% US style
+%  \\[1em]% UK style
+  \@date\\[2em]% US informal style: "January 1, 1900"; UK formal style: "01/01/1900"
+  % opening
```
Done!

jalopezg-git commented:
> > For this to work, I basically hacked an ad-hoc replacement for some of the skill matrix macros (see attached file). I can try to make moderncvskillmatrix officially work in this PR 🙂 - if I don't find the time, are you okay if we leave it as a follow-up?
> > Attachment: [moderncvskillmatrixtweaks.tex.gz](https://github.com/moderncv/moderncv/files/15133715/moderncvskillmatrixtweaks.tex.gz)
> 
> Thanks for your attachment. First I just thought we could try to support the examples in the `templates.tex`. But if there is more work to be done to integrate the skill matrix maybe we can do that in a separate task, yes.

No worries, I think I can look into it and include it in this PR :slightly_smiling_face:.

jalopezg-git commented:
Thanks for the review, @LLdaniel :+1:!  Hopefully this should be ready for merging now!

I have addressed all your comments.  Also, I have rebased atop the current `master` state; hope it doesn't make the final review harder.  Probably you are mostly interested in taking a look to the changes in the last 5 commits.

LLdaniel reviewed (commented):
Great! Thank you very much.
I also want to merge this PR soon and wanted to reach out to your opinion regarding the following problem (see comment below). 

Maybe we can decide together what is the best option to go on. In either way I would approve this PR to at least get it merged and work with the new style from there on.

jalopezg-git commented on moderncvstylecontemporary.sty:
```diff
@@ -0,0 +1,55 @@
+%\fi
+
+% symbols
+\moderncvicons{marvosym}
```
Hi @LLdaniel,

> 2\. Taking the change here and risiking a bad color scheme because the symbols will be all grey then. I also will create a follow up issue to address the fact that it is difficult at the moment to change the icon color at different positions in the document. _The workaround would be to use grey symbols everywhere which does not exactly match the intended color scheme for "contemporary"._

Actually, I switched to `\moderncvicons{symbols}` and slightly changed `common/colors.tex` to define `default-socialicon-color` only if a previous definition is not available.
I also applied some some other minor changes to `moderncvstylecontemporary.sty` that were required after merging #151.

There seems to be a small issue regarding the coloring of the Signal icon, but seems totally unrelated to this PR.  Maybe doable in a different one (?).

Have a nice rest of the day!

LLdaniel replied:
Hi @jalopezg-git, 
thanks a lot for your work!
> Actually, I switched to \moderncvicons{symbols} and slightly changed common/colors.tex to define default-socialicon-color only if a previous definition is not available. Now icons match the intended color scheme for 'contemporary'.

Great that the symbols are white now and match the overall style. One minor issue appears now that the symbols in the cover letter are also white. But I open a issue for that, because at the moment it is not possible to solve that for two locations at a time. Also for me the PR seems "big" enough^^

>I also applied some some other minor changes to moderncvstylecontemporary.sty that were required after merging https://github.com/moderncv/moderncv/pull/151.

Thanks in retrospective this was not the cleverest move from my side to integrate that PR first. At a first glance it seemed to me that we need that, but in fact it was more work for you... Sry.

>There seems to be a small issue regarding the coloring of the Signal icon, but seems totally unrelated to this PR. Maybe doable in a different one (?).

Yes this is more a matter of which kind of icon is picked up. Fontawesome 5 does not have a Signal icon and there is currently no Matrix icon as well. That is the reason why they are different. But this was **not** introduced with your PR, you are right. 

>Totally unrelated to the above, but 'contemporary' looks better with a much thinner header. Perhaps we could provide a template.tex that has most of the uncommon \social commented out + a suggestion to uncomment if needed.

Yeah, using the full list of icons is a bit overkill and hopefully nobody needs that in his/her CV. I know that it distorts the style this also happens for example with `\moderncvstyle[details]{classic}`.
For me it is ok to use show the full set of capable social icons with the knowledge that hopefully the full set is not used. But it would be also okay if we try to shorten the list and leave other icons only in the comments of `template.tex`.

jalopezg-git replied:
Hi @LLdaniel,

First of all, many thanks for quick the review :+1:!
 
> Great that the symbols are white now and match the overall style. One minor issue appears now that the symbols in the cover letter are also white. But I open a issue for that, because at the moment it is not possible to solve that for two locations at a time. Also for me the PR seems "big" enough^^

Oops, I did not notice when working on the patch!  Okay, let's address that in a follow-up PR!

> Thanks in retrospective this was not the cleverest move from my side to integrate that PR first. At a first glance it seemed to me that we need that, but in fact it was more work for you... Sry.

No worries!

> Yes this is more a matter of which kind of icon is picked up. Fontawesome 5 does not have a Signal icon and there is currently no Matrix icon as well. That is the reason why they are different. But this was **not** introduced with your PR, you are right.

Great!  I'll merge now and we can address this minor issues separately :slightly_smiling_face:.

> Yeah, using the full list of icons is a bit overkill and hopefully nobody needs that in his/her CV. I know that it distorts the style this also happens for example with `\moderncvstyle[details]{classic}`. For me it is ok to use show the full set of capable social icons with the knowledge that hopefully the full set is not used. But it would be also okay if we try to shorten the list and leave other icons only in the comments of `template.tex`.

Let's merge as-is and then we can change to a thinner header version, if need be.

LLdaniel reviewed (approved):
Awesome, @jalopezg-git! 
In my opinion we are ready to merge your new style. 
Thanks a lot!

jalopezg-git commented:
Hi @LLdaniel,

First of all, many thanks for quick the review :+1:!
 
> Great that the symbols are white now and match the overall style. One minor issue appears now that the symbols in the cover letter are also white. But I open a issue for that, because at the moment it is not possible to solve that for two locations at a time. Also for me the PR seems "big" enough^^

Oops, I did not notice when working on the patch!  Okay, let's address that in a follow-up PR!

> Thanks in retrospective this was not the cleverest move from my side to integrate that PR first. At a first glance it seemed to me that we need that, but in fact it was more work for you... Sry.

No worries!

> Yes this is more a matter of which kind of icon is picked up. Fontawesome 5 does not have a Signal icon and there is currently no Matrix icon as well. That is the reason why they are different. But this was **not** introduced with your PR, you are right.

Great!  I'll merge now and we can address this minor issues separately :slightly_smiling_face:.

> Yeah, using the full list of icons is a bit overkill and hopefully nobody needs that in his/her CV. I know that it distorts the style this also happens for example with `\moderncvstyle[details]{classic}`. For me it is ok to use show the full set of capable social icons with the knowledge that hopefully the full set is not used. But it would be also okay if we try to shorten the list and leave other icons only in the comments of `template.tex`.

Let's merge as-is and then we can change to a thinner header version, if need be.

jalopezg-git commented:
> Awesome, @jalopezg-git! In my opinion we are ready to merge your new style. Thanks a lot! Are you allowed to merge it or should I do it on behalf of you?

Yeah, I was allowed to merge it!  Many thanks for your reviews!
